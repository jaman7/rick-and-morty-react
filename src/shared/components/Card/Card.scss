@use "sass:color";
@import "@scss/extend";
@import "@scss/variables";
@import "@scss/mixins";

.card {
  display: flex;
  flex-direction: column;
  background-color: var(--main-bg-white);
  border-radius: 0.5rem;
  box-shadow: 0 4px 12px rgb(0 0 0 / 15%);
  overflow: hidden;
  transition:
    transform 1s cubic-bezier(0.36, 0.17, 0.12, 1.23),
    box-shadow 0.2s ease-in-out;
  justify-content: space-between;
  transform: scale(1);
  will-change: transform;

  &:hover {
    transform: perspective(800px) rotateY(6deg) scale(1.02);
    transform-style: preserve-3d;

    @extend %shared-box-shadow;

    @include mq(maxlg) {
      transform: perspective(600px) rotateY(4deg) scale(1.01);
    }

    @include mq(maxsm) {
      transform: perspective(600px) rotateY(2deg) scale(1.005);
    }
  }

  &-container-link {
    position: relative;
    flex: 1;

    &:hover {
      & .card-overlay-box {
        opacity: 0.85;
        transform: translate(0, 0) rotate(0) skew(0);
      }

      & .card-overlay-box-text {
        opacity: 1;
        transform: translate(-50%, -50%) rotate(0) skew(0);
        transition-delay: 0.25s;
      }
    }
  }

  &-body {
    padding: 1rem;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    gap: 0.5rem;
    flex: 1;
  }

  &-title {
    font-size: 1.125rem;
    font-weight: 600;
    color: var(--main-text-primary);
    text-align: center;
    margin-bottom: 0.5rem;

    @extend %text-shadow-blue;
  }

  &-footer {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: auto;
    font-size: 0.875rem;
    color: var(--main-text-secondary);

    & .link-item {
      color: var(--main-text-link);
      padding: 0.5rem;
      font-weight: 700;

      &:hover {
        color: var(--main-text-orange);
      }

      &.active {
        color: var(--main-text-orange);
      }
    }
  }

  &-link {
    color: var(--main-text-link);
    background-color: var(--main-bg-orange);
    border-radius: 0.5rem;
    padding: 0.5rem;
    font-weight: 700;
    transition:
      color 250ms ease-in-out,
      background-color 250ms ease-in-out;

    @extend %shared-box-shadow;

    &:hover {
      color: var(--main-text-white);
      background-color: color.scale($colorBlue, $lightness: 30%);
    }
  }

  &-overlay-box {
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 10;
    top: 0;
    left: 0;
    background-color: color.scale($colorOrange, $lightness: 70%);
    box-shadow: inset 2px 2px 12px 0 rgb(0 0 0 / 20%);
    opacity: 0;
    transform: translate(-100%, -100%) rotate(-30deg) skew(-30deg);
    transition:
      transform 0.5s ease-in-out,
      opacity 0.5s ease-in-out;
  }

  &-overlay-box-text {
    font-weight: 700;
    position: absolute;
    opacity: 0;
    color: var(--main-text-primary);

    @include fluid-type(1.8rem, 2.5rem);

    z-index: 11;
    top: 50%;
    left: 50%;
    width: 100%;
    text-align: center;
    transform: translate(-90%, -50%) rotateX(45deg) skewX(20deg);
    transition:
      transform 0.5s ease-in-out,
      opacity 0.5s ease-in-out;

    @extend %text-shadow-blue;
  }
}
