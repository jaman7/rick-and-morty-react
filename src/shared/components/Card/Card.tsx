import { useEffect, useState, ReactNode, CSSProperties } from 'react';
import { motion } from 'framer-motion';
import LazyImage from '../LazyImage/LazyImage';
import useResizeObserver from '@/hooks/useResizeObserver';
import classNames from 'classnames';
import './Card.scss';
import { NavLink } from 'react-router-dom';

interface CardProps {
  name: string;
  isEffect?: boolean;
  className?: string;
  image?: string | null;
  onResize?: (width: number, height: number) => void;
  link?: string;
  linkTarget?: '_blank' | '_self' | '_parent' | '_top' | '';
  clickable?: boolean;
  showFooterLink?: boolean;
  children?: ReactNode;
  style?: CSSProperties;
}

const Card = ({
  name = '',
  isEffect = false,
  className = '',
  image = '',
  onResize,
  showFooterLink = false,
  clickable = false,
  link = '',
  linkTarget = '',
  children,
  style = {},
  ...rest
}: CardProps) => {
  const anchorObserver = useResizeObserver<HTMLAnchorElement>();
  const divObserver = useResizeObserver<HTMLDivElement>();

  const isLink = clickable && !!link;
  const { size } = clickable && isLink ? anchorObserver : divObserver;
  const resizeRef = clickable && isLink ? anchorObserver : divObserver;
  const [prevSize, setPrevSize] = useState(resizeRef.size);

  useEffect(() => {
    if (size.width !== prevSize.width || size.height !== prevSize.height) {
      setPrevSize(size);
      onResize?.(size.width, size.height);
    }
  }, [size.width, size.height]);

  const variants = {
    initial: { opacity: 0, y: 10 },
    animate: { opacity: 1, y: 0 },
  };

  const content = (
    <>
      {image && <LazyImage src={image} alt={name} />}
      <div className="card-body">
        {name && (
          <div className="card-title" id={`${name}-title`}>
            {name}
          </div>
        )}
        {children}
        {showFooterLink && link && !isEffect && (
          <div className="d-block mt-2">
            <NavLink to={link} target={linkTarget || undefined} className="card-link">
              View more
            </NavLink>
          </div>
        )}
      </div>
    </>
  );

  const overlay = isEffect && (
    <>
      <div className="card-overlay-box" aria-hidden="true" />
      <div className="card-overlay-box-text" aria-hidden="false" aria-label="Details">
        Details
      </div>
    </>
  );

  const innerContent = (
    <>
      {content}
      {overlay}
    </>
  );

  const id = `card-title-${name.replace(/\s+/g, '-').toLowerCase()}`;
  const isExternal = linkTarget === '_blank';
  const linkProps = {
    to: link,
    target: linkTarget || undefined,
    rel: isExternal ? 'noopener noreferrer' : undefined,
  };

  return (
    <motion.div
      ref={divObserver.ref}
      role="group"
      aria-labelledby={id}
      className={classNames('card', className, { hoverable: isEffect })}
      style={style}
      variants={variants}
      transition={{ duration: 0.3 }}
      {...rest}
    >
      {clickable && link ? (
        <NavLink className="card-container-link flex-column space-between" {...linkProps} {...rest}>
          {innerContent}
        </NavLink>
      ) : (
        innerContent
      )}
    </motion.div>
  );
};

export default Card;
