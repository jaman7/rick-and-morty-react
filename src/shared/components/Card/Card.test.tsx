import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { act, render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import Card from './Card';

describe('Card', () => {
  const observe = vi.fn();
  const unobserve = vi.fn();
  const disconnect = vi.fn();

  beforeEach(() => {
    // @ts-ignore: mock global ResizeObserver
    global.ResizeObserver = vi.fn(() => ({
      observe,
      unobserve,
      disconnect,
    }));
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders title and children', () => {
    render(
      <Card name="Rick Card">
        <p>Child content</p>
      </Card>
    );
    expect(screen.getByText('Rick Card')).toBeInTheDocument();
    expect(screen.getByText('Child content')).toBeInTheDocument();
  });

  it('applies hoverable class if hoverable is true', () => {
    const { container } = render(<Card name="Hover" isEffect />);
    const cardDiv = container.querySelector('.card');
    expect(cardDiv?.classList.contains('hoverable')).toBe(true);
  });

  it('calls onResize when size changes', async () => {
    const onResize = vi.fn();

    render(<Card name="Resizable" onResize={onResize} />);

    const mockEntry = [
      {
        contentRect: { width: 200, height: 100 },
      },
    ];

    const callback = (global.ResizeObserver as unknown as ReturnType<typeof vi.fn>).mock.calls[0][0];

    await act(() => {
      callback(mockEntry);
    });

    expect(onResize).toHaveBeenCalledWith(200, 100);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Card name="Snapshot" isEffect className="custom-class">
        <span>Snapshot test</span>
      </Card>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
