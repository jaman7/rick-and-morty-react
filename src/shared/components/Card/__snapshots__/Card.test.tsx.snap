// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="card-title-snapshot"
    class="card custom-class hoverable"
    role="group"
  >
    <div
      class="card-body"
    >
      <div
        class="card-title"
        id="Snapshot-title"
      >
        Snapshot
      </div>
      <span>
        Snapshot test
      </span>
    </div>
    <div
      aria-hidden="true"
      class="card-overlay-box"
    />
    <div
      aria-hidden="false"
      aria-label="Details"
      class="card-overlay-box-text"
    >
      Details
    </div>
  </div>
</DocumentFragment>
`;
